pkg/sysregistriesv2/system_registries_v2.go

```diff
@@ -130,17 +146,19 @@ type PullSource struct {
 // reference.
 func (r *Registry) PullSourcesFromReference(ref reference.Named) ([]PullSource, error) {
 	var endpoints []Endpoint
-
-	if r.MirrorByDigestOnly {
-		// Only use mirrors when the reference is a digest one.
-		if _, isDigested := ref.(reference.Canonical); isDigested {
-			endpoints = append(r.Mirrors, r.Endpoint)
+	_, isDigested := ref.(reference.Canonical)
+	for _, mirror := range r.Mirrors {
+		if isDigested {
+			if !mirror.TagOnly && !r.MirrorByTagOnly || mirror.DigestOnly {
```

Comment:
If the model isn’t simplified, and these conditions remain (see elsewhere): Please use at least parentheses to be explicit about the operator priority / intent here.
Maybe even something like
```go
if mirror.DigestOnly // explicitly opted in
  || (!mirror.TagOnly && !r.MirrorByTagOnly) { // not opted out
```

Admittedly the `(!A && !B) || C` and `!A && (!B | C)` expressions differ only in the forbidden case of `mirror.TagOnly && mirror.DigestOnly`, but I ended up actually making a table of all 8 possible inputs to the two expressions and evaluating the two expressions to figure that out (admittedly it can also be done symbolically) — either way I think that’s a bit too much to ask of anyone modifying this code in the future, possibly in a hurry.

---

Same in the other condition.